### PR TITLE
[WIP] Quagga: Add BGP to UI

### PR DESCRIPTION
**Pull request by mimugmail**
Ok, so I've changed all the stuff, but still the bgpd.conf is empty and there's no error when trying to enable it.

### CODE REVIEW

**Comment by fabianfrz** [comment]
@fichtner this will definitely be a squashed commit ;)

**Comment by mimugmail** [comment]
@fabianfrz Found the error. I copied OspfsettingsController but referenced to BgpController. Renamed the file an now bgpd.conf gets created.

**Comment by fabianfrz** [comment]
sounds good. now you can start adding the advanced stuff like network configuration etc. this is the large code in the ospfsettingscontroller (for two sub-models in an 1:n relation)

File: net/quagga/src/opnsense/mvc/app/controllers/OPNsense/Quagga/forms/bgp.xml [hunk 1]
```diff
@@ -12,6 +12,14 @@
         <hint>Your AS Number here</hint>
     </field>
     <field>
+        <id>bgp.networks</id>
+        <label>Network</label>
+        <style>tokenize</style>
+        <type>select_multiple</type>
+        <allownew>true</allownew>
+        <help>Select the network to advertise, you have to set a Null route via zebra or a static route.</help>
```
Comment (fabianfrz) [inline net/quagga/src/opnsense/mvc/app/controllers/OPNsense/Quagga/forms/bgp.xml]: At the moment, the zebra configuration is empty (only an on / off tick in general). I don't want to change that as this is the job of the system configuration pages (setting static routes).

Comment (mimugmail) [inline net/quagga/src/opnsense/mvc/app/controllers/OPNsense/Quagga/forms/bgp.xml]: Hm, yes you're right. In my daily work I use Null0 via Zebra, but OPNsense can set Null routes too, so I'll update the help.

**Comment by mimugmail** [comment]
@fabianfrz With copying your OSPF stuff I added the functionality to add multiple neighbors for BGP, but I'm unsure about the systax for bgpd.conf generation. I took the cline stuff you used in ospfd.conf, but with bgp there is not always a "directive" "modelname".

A typical neighbor config looks like this:

 neighbor <address> remote-as <remoteas>
 neighbor <address> update-source <iface>
 neighbor <address> next-hop-self

Could you give me a hint how to solve this? 

Also ATM the interface config is set to manual, I'll change this when you have spoken to @fichtner regarding the interface detection.

Thanks!
Michael

**Comment by fabianfrz** [comment]
you ave an address model with the fields remoteas, iface etc. and next-hop-self may be an boolean.

So I would add it by first checking the address with an if. the other lines also with an if  and print them if they are valid like it is done everywhere else.
https://github.com/opnsense/plugins/blob/master/net/quagga/src/opnsense/service/templates/OPNsense/Quagga/ospfd.conf#L29-L32

The {{ value }} should be set to the field you are using. Because there are many fields looking the same, I created a macro for OSPF.

**Comment by mimugmail** [comment]
@fabianfrz This was the hint I needed, thanks! 
Now I have a working setup with remote-as. I'll try the other boolean stuff later.

**Comment by mimugmail** [comment]
@fabianfrz OK please review, I'm finished now. 
I'm able to add und remove peer's, set remote as and next-hop-self and update-source.

With this stage you can run basic eBGP and iBGP. 

There are hundreds of options missing like route-maps etc. but for basic BGP this runs fine.

File: net/quagga/src/opnsense/service/templates/OPNsense/Quagga/bgpd.conf [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+{%   for network in OPNsense.quagga.bgp.networks.split(',') %}
+ network {{ network }}
+{%   endfor %}
+{%  endif %}
+{% if helpers.exists('OPNsense.quagga.bgp.neighbors.neighbor') %}
+{%   for neighbor in helpers.toList('OPNsense.quagga.bgp.neighbors.neighbor') %}
+{%     if neighbor.enabled == '1' %}
+ neighbor {{ neighbor.address }} remote-as {{ neighbor.remoteas }} 
+{%     endif %}
+{%     if neighbor.enabled == '1' and neighbor.updatesource != '' %}
```
Comment (fabianfrz) [inline net/quagga/src/opnsense/service/templates/OPNsense/Quagga/bgpd.conf]: This is a field which is not required - replace `neighbor.enabled == '1'` by an existence check - also at the other lines later

File: net/quagga/src/opnsense/mvc/app/views/OPNsense/Quagga/bgp.volt [hunk 3]
```diff
@@ -1 +1,103 @@
-{{ partial("layout_partials/base_form",['fields':generalForm,'id':'frm_ospf_settings'])}}
+{#
+
+OPNsense® is Copyright © 2014 – 2017 by Deciso B.V.
+This file is Copyright © 2017 by Fabian Franz
```
Comment (fabianfrz) [inline net/quagga/src/opnsense/mvc/app/views/OPNsense/Quagga/bgp.volt]: really not your copyright?

File: net/quagga/src/opnsense/mvc/app/controllers/OPNsense/Quagga/Api/BgpController.php [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+            if ($valMsgs->count() == 0) {
+                $mdlBGP->serializeToConfig();
+                Config::getInstance()->save();
+                $result["result"] = "saved";
+            }
+        }
+        return $result;
+    }
+
+///////////////////////////////////////////////////////////////////// TO BE REVIEWED
```
Comment (fabianfrz) [inline net/quagga/src/opnsense/mvc/app/controllers/OPNsense/Quagga/Api/BgpController.php]: remove this line

File: net/quagga/src/opnsense/mvc/app/controllers/OPNsense/Quagga/Api/BgpController.php [hunk 5]
```diff
@@ -0,0 +1,194 @@
+<?php
+namespace OPNsense\Quagga\Api;
+use \OPNsense\Base\ApiControllerBase;
+use \OPNsense\Quagga\BGP;
+use \OPNsense\Core\Config;
+use \OPNsense\Base\ApiMutableModelControllerBase;
+use \OPNsense\Base\UIModelGrid;
+/**
+ *    Copyright (C) 2015 - 2017 Deciso B.V.
+ *    Copyright (C) 2017 Fabian Franz
```
Comment (fabianfrz) [inline net/quagga/src/opnsense/mvc/app/controllers/OPNsense/Quagga/Api/BgpController.php]: Not your copyright?

File: net/quagga/src/opnsense/mvc/app/models/OPNsense/Quagga/BGP.xml [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+                        <updatesource type="TextField">
+                                <default></default>
+                                <Required>N</Required>
+                        </updatesource>
+						<nexthopself type="BooleanField">
+                                <default>0</default>
+                                <Required>N</Required>
+                        </nexthopself>
+						<defaultoriginate type="BooleanField">
+                                <default>0</default>
```
Comment (fabianfrz) [inline net/quagga/src/opnsense/mvc/app/models/OPNsense/Quagga/BGP.xml]: same here

File: net/quagga/src/opnsense/mvc/app/models/OPNsense/Quagga/BGP.xml [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+                        </address>
+                        <remoteas type="TextField">
+                                <default></default>
+                                <Required>Y</Required>
+                        </remoteas>
+                        <updatesource type="TextField">
+                                <default></default>
+                                <Required>N</Required>
+                        </updatesource>
+						<nexthopself type="BooleanField">
```
Comment (mimugmail) [inline net/quagga/src/opnsense/mvc/app/models/OPNsense/Quagga/BGP.xml]: corrected

File: net/quagga/src/opnsense/mvc/app/views/OPNsense/Quagga/bgp.volt (hunk 3 above)
Comment (mimugmail) [inline net/quagga/src/opnsense/mvc/app/views/OPNsense/Quagga/bgp.volt]: I copied your file and replaced ospf with bgp, it's yours :)
I added a noted with my name

File: net/quagga/src/opnsense/mvc/app/views/OPNsense/Quagga/bgp.volt [hunk 8]
```diff
@@ -1 +1,103 @@
...
+
+<table id="grid-neighbors" class="table table-responsive" data-editDialog="DialogEditBGPNeighbor">
+<thead>
+            <tr>
+                <th data-column-id="enabled" data-type="string" data-formatter="rowtoggle">{{ lang._('Enabled') }}</th>
+                <th data-column-id="address" data-type="string" data-visible="true">{{ lang._('Neighbor Address') }}</th>
+                <th data-column-id="remoteas" data-type="string" data-visible="true">{{ lang._('Remote AS') }}</th>
+                <th data-column-id="updatesource" data-type="string" data-visible="true">{{ lang._('Update Source Address') }}</th>
+		<th data-column-id="nexthopself" data-type="string" data-formatter="rowtoggle">{{ lang._('Next Hop Self') }}</th>
+		<th data-column-id="defaultoriginate" data-type="string" data-formatter="rowtoggle">{{ lang._('Default Originate') }}</th>
```
Comment (mimugmail) [inline net/quagga/src/opnsense/mvc/app/views/OPNsense/Quagga/bgp.volt]: For me it looks ok?

File: net/quagga/src/opnsense/mvc/app/views/OPNsense/Quagga/bgp.volt [hunk 9]
```diff
@@ -1 +1,103 @@
...
+            </tr>
+            </thead>
+            <tbody>
+            </tbody>
+            <tfoot>
+            <tr>
+                <td colspan="5"></td>
+                <td>
+                    <button data-action="add" type="button" class="btn btn-xs btn-default"><span class="fa fa-plus"></span></button>
+                    <!-- <button data-action="deleteSelected" type="button" class="btn btn-xs btn-default"><span class="fa fa-trash-o"></span></button> -->
```
Comment (mimugmail) [inline net/quagga/src/opnsense/mvc/app/views/OPNsense/Quagga/bgp.volt]: I'd let it there if it's ok for you

File: net/quagga/src/opnsense/mvc/app/views/OPNsense/Quagga/bgp.volt (hunk 3 above)
Comment (fichtner) [inline net/quagga/src/opnsense/mvc/app/views/OPNsense/Quagga/bgp.volt]: files are copied, your own copyrights are added, no additions that weren't in there before. @fabianfrz should must his copyrights on his own

File: net/quagga/src/opnsense/service/templates/OPNsense/Quagga/bgpd.conf (hunk 2 above)
Comment (mimugmail) [inline net/quagga/src/opnsense/service/templates/OPNsense/Quagga/bgpd.conf]: If I don't check for neighbor enabled these line will stay if I disable a neighbor. I'll try to put the whole thing under an if as you stated before

File: net/quagga/src/opnsense/mvc/app/controllers/OPNsense/Quagga/Api/BgpController.php (hunk 5 above)
Comment (mimugmail) [inline net/quagga/src/opnsense/mvc/app/controllers/OPNsense/Quagga/Api/BgpController.php]: Same as above, it's yours!

File: net/quagga/src/opnsense/mvc/app/views/OPNsense/Quagga/bgp.volt (hunk 8 above)
Comment (fichtner) [inline net/quagga/src/opnsense/mvc/app/views/OPNsense/Quagga/bgp.volt]: tabs vs. spaces, indent is PSR style, 4 spaces per indent

but we have a style fixer for this too

File: net/quagga/src/opnsense/mvc/app/controllers/OPNsense/Quagga/Api/BgpController.php (hunk 4 above)
Comment (mimugmail) [inline net/quagga/src/opnsense/mvc/app/controllers/OPNsense/Quagga/Api/BgpController.php]: Ok, it's gone and I added some newlines before the functions

File: net/quagga/src/opnsense/mvc/app/controllers/OPNsense/Quagga/BgpController.php [hunk 10]
```diff
@@ -2,6 +2,7 @@
 namespace OPNsense\Quagga;
 /*
     Copyright (C) 2017 Fabian Franz
+    Modified for BGP functionality by Michael Muenz
```
Comment (fichtner) [inline net/quagga/src/opnsense/mvc/app/controllers/OPNsense/Quagga/BgpController.php]: It should really read the following:

Copyright (C) 2017 Michael Muenz

don't be shy :)

File: net/quagga/src/opnsense/mvc/app/models/OPNsense/Quagga/BGP.xml (hunk 6 above)
Comment (mimugmail) [inline net/quagga/src/opnsense/mvc/app/models/OPNsense/Quagga/BGP.xml]: Already fixed

File: net/quagga/src/opnsense/mvc/app/views/OPNsense/Quagga/bgp.volt (hunk 8 above)
Comment (mimugmail) [inline net/quagga/src/opnsense/mvc/app/views/OPNsense/Quagga/bgp.volt]: Hm, within github UI it's ok. So I'll leave it and you'll fix it?

File: net/quagga/src/opnsense/service/templates/OPNsense/Quagga/bgpd.conf [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+{% if helpers.exists('OPNsense.quagga.bgp.asnumber') and OPNsense.quagga.bgp.asnumber != '' %}
+router bgp {{ OPNsense.quagga.bgp.asnumber }}
+{% if helpers.exists('OPNsense.quagga.bgp.networks') %}
+{%   for network in OPNsense.quagga.bgp.networks.split(',') %}
+ network {{ network }}
+{%   endfor %}
+{%  endif %}
+{% if helpers.exists('OPNsense.quagga.bgp.neighbors.neighbor') %}
+{%   for neighbor in helpers.toList('OPNsense.quagga.bgp.neighbors.neighbor') %}
+{%     if neighbor.enabled == '1' %}
```
Comment (mimugmail) [inline net/quagga/src/opnsense/service/templates/OPNsense/Quagga/bgpd.conf]: Ok I put all under an if, seems to work fine, thanks for the hint!

**Comment by mimugmail** [comment]
@fabianfrz @fichtner Ok I'm done with the reviews so far, thanks for the fast review!

File: net/quagga/src/opnsense/mvc/app/views/OPNsense/Quagga/bgp.volt (hunk 8 above)
Comment (fabianfrz) [inline net/quagga/src/opnsense/mvc/app/views/OPNsense/Quagga/bgp.volt]: You mixed spaces and tabs. I replaced the tabs in those two lines with the same amount of spaces like the other lines. looks better now. In general, we do not use tabs in the source.

File: net/quagga/src/opnsense/mvc/app/views/OPNsense/Quagga/bgp.volt (hunk 9 above)
Comment (fabianfrz) [inline net/quagga/src/opnsense/mvc/app/views/OPNsense/Quagga/bgp.volt]: no problem.

**Comment by mimugmail** [comment]
@fabianfrz Stupid question: Do have to edit the indent stuff on my own or will your commit merged too?

**Comment by fabianfrz** [comment]
Everything will be merged together as it is on your branch.

**Comment by fabianfrz** [comment]
@mimugmail is is probably only the template that needs changes (maybe whitespace stuff too). I think we can merge it after this changes are done. I am sure @fichtner could do the whitespace cleanup.

**Comment by fabianfrz** [comment]
I close this for #121

**Comment by fabianfrz** [comment]
@mimugmail your changes are merged now.

**Comment by fabianfrz** [comment]
Please delete your changes from `master` now or create a new clone so you can fast forward your `master` branch.

**Pull request merged**